crates/ink/ir/src/ir/item_impl/constructor.rs

```diff
@@ -92,7 +92,34 @@ impl Constructor {
         }) if path.is_ident("Self"))
     }
 
-    /// Ensures that the return type of the ink! constructor is `Self`.
+    /// Returns `true` if the given type is `Result<Self>`.
+    fn type_is_result_self_val(ty: &syn::Type) -> bool {
+        let res = matches!(ty, syn::Type::Path(syn::TypePath {
+            qself: None,
+            path
+        }) if Self::is_result_path(path));
+        res
+    }
+
+    /// Helper functions that evaluates that the return type is `Result`
+    /// and the `Ok` variant is `Self` type
+    fn is_result_path(path: &syn::Path) -> bool {
+        if path.leading_colon.is_none()
+            && path.segments.last().is_some()
+            && path.segments.last().unwrap().ident == "Result"
+        {
+            if let syn::PathArguments::AngleBracketed(angle_args) =
+                &path.segments.last().unwrap().arguments
+            {
+                if let Some(syn::GenericArgument::Type(ty)) = angle_args.args.first() {
+                    return Self::type_is_self_val(ty)
+                }
+            }
+        }
+        false
+    }
```

Comment:
This looks as if you are checking the return type of an ink! constructor by checking if the identifier of the returned value type is equal to `"Result"`. Therefore you use syntactical checks instead of making use of the powerful Rust type system.

This violates the rules of proc. macro hygiene and is probably also the reason why this implementation depends on the `type Result` alias that @athei has already mentioned.

Let me be clear about this: ink! or any other Rust proc. macro that wants to be taken seriously should never make use of syntactical analysis for thing like this and instead should _always_ make use of the powerful Rust type system whenever it is possible and it is very well possible here and we discussed about this some time ago in chat.

If you need more elaboration please feel free to ask me and I will happily provide you with answers.

Reply:
While I agree with the philosophy, it is a little bit tricky in this case to do:
* We have `is_result_type!` macro that check `Result` based on type. However, it sits in another crate, and I will need to import the whole crate for a single macro.
* Even though this might not be a problem. The real problem lies in `Self`. We still need to extract `Ok` variant and **syntacticly** check for `Self`. Otherwise, if we want to utilise the type system in this case too, we need to know the storage type and then compare it to the `Ok` variant return type. However, my understanding is that `ir` crate sits in isolation from the global `ink` crate and context and makes no assumptions what happens "*beyond the syntax*", it solely works with token stream. 
* Hence, that led me to thinking that we only want to ensure that `Result` keyword with `Self` are present in the token stream. The real type checks happen later down the pipeline when we generate dispatchable messages and metadata

Reply:
Syntactical checks are red flags for proc. macros generally. So yes.

Reply:
One possible avenue of investigation to replace the syntactica checkl is the use of [`DispatchableConstructorInfo::Output`](https://github.com/paritytech/ink/blob/109536d43abab7d6eed5000c4b208b74c49aa106/crates/ink/src/reflect/dispatch.rs#L335) to constrain the return type to either be `Self::Storage` or `Result<Self::Storage>`. Might require an additional trait with auto impl for the `Result`, or maybe there is something already there we can reuse.

Reply:
> One possible avenue of investigation to replace the syntactica checkl is the use of [`DispatchableConstructorInfo::Output`](https://github.com/paritytech/ink/blob/109536d43abab7d6eed5000c4b208b74c49aa106/crates/ink/src/reflect/dispatch.rs#L335) to constrain the return type to either be `Self::Storage` or `Result<Self::Storage>`. Might require an additional trait with auto impl for the `Result`, or maybe there is something already there we can reuse.

That might be a good option. I was thinking of using an enum with to variants. `Raw(Storage)` and `Result(Storage)`